fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/partition/PartitionSelector.java

```diff
@@ -247,6 +264,99 @@ private static List<Long> getPartitionIdsByExpr(ConnectContext context,
+            }
+        }
+        for (ColumnRefOperator columnRef : columnRefOperatorList) {
+            Column column = olapTable.getColumn(columnRef.getName());
+            if (column != null) {
+                if (columnNameToColRefMap.containsKey(column.getName())) {
+                    columnRef.setType(column.getType());
+                }
+                columnNameToColRefMap.put(column.getName(), columnRef);
+            }
+        }
+        Function<SlotRef, ColumnRefOperator> slotRefResolver = (slot) -> {
+            return columnNameToColRefMap.get(slot.getColumnName());
+        };
+        // The GeneratedColumn doesn't have the correct type info, let's help it
+        Consumer<SlotRef> slotRefConsumer = (slot) -> {
+            ColumnRefOperator ref = columnNameToColRefMap.get(slot.getColumnName());
+            if (ref != null) {
+                slot.setType(ref.getType());
+            }
+        };
+
+        // Build a map of c1 -> c3, in which c3=fn(c1)
+        Map<ColumnRefOperator, Pair<ColumnRefOperator, ScalarOperator>> refedGeneratedColumnMap = Maps.newHashMap();
+        for (ColumnRefOperator partitionColumn : partitionColumnRefs) {
+            Column column = olapTable.getColumn(partitionColumn.getName());
+            if (column != null && column.isGeneratedColumn()) {
+                Expr generatedExpr = column.getGeneratedColumnExpr(olapTable.getBaseSchema());
+                ExpressionAnalyzer.analyzeExpressionResolveSlot(generatedExpr, ConnectContext.get(), slotRefConsumer);
+                ScalarOperator call =
+                        SqlToScalarOperatorTranslator.translateWithSlotRef(generatedExpr, slotRefResolver);
+
+                if (call instanceof CallOperator &&
+                        OperatorFunctionChecker.onlyContainMonotonicFunctions((CallOperator) call).first) {
+                    columnRefOperatorList = Utils.extractColumnRef(call);
+
+                    for (ColumnRefOperator ref : columnRefOperatorList) {
+                        refedGeneratedColumnMap.put(ref, Pair.create(partitionColumn, call));
+                    }
+                }
+            }
+        }
+
+        // No GeneratedColumn with partition column
+        if (refedGeneratedColumnMap.isEmpty()) {
+            return scalarOperator;
+        }
+
+        ScalarOperatorVisitor<ScalarOperator, Void> visitor = new ScalarOperatorVisitor<>() {
+            @Override
+            public ScalarOperator visit(ScalarOperator scalarOperator, Void context) {
+                List<ColumnRefOperator> columnRefOperatorList = Utils.extractColumnRef(scalarOperator);
+                if (checkDeduceConjunct(partitionColumnRefs, scalarOperator, columnRefOperatorList)) {
+                    ColumnRefOperator referenced = columnRefOperatorList.get(0);
+                    Pair<ColumnRefOperator, ScalarOperator> pair = refedGeneratedColumnMap.get(referenced);
+                    if (pair != null) {
```

Comment:
Can you rebase your codes?  I doubt there may some conflicts with this.

Reply:
`buildDeducedConjunct` is safe for `str2date` or other partition expressions since it may be different types for input and output types of `str2date`?  maybe we can add more tests about this.

Reply:
add some str2date test cases